students/FlorentinPopescu/12_UnitTesting_of_MailroomPart4_L6_FP.py

```diff
@@ -0,0 +1,170 @@
+        del mailroom.donors["Florentin"]
+
+#------------------------------------------------
+    def test_write_letter(self):
+        Florentin = io.StringIO()
+        
+        with redirect_stdout(Florentin):
+            mailroom.write_letter({"Florentin": [20]})
+        
+        out = Florentin.getvalue()
+        self.assertEqual("Prepared letter for 'Florentin'.\n", out)
+        self.assertFalse(os.path.isdir("./Florentin.txt"))
+        self.assertTrue(os.path.exists("Florentin.txt"), "Florentin.txt")
+        
+        with open("Florentin.txt", "r") as f:
+            text = f.read()
+        self.assertEqual("\nDear Florentin, \nThank you for you donation of $20.00 to our charity. \nSincerely.", text)
+  
+#------------------------------------------------
+    def test_switch_directories(self):
+        current_directory = os.getcwd()
+        new_directory = "repository"
+        final_directory = os.path.join(current_directory, new_directory)
+        
+        if not os.path.exists(final_directory):
+           os.makedirs(final_directory)
+        
+        mailroom.switch_directories(mailroom.donors, current_directory, new_directory)
+        self.assertEqual(len(mailroom.donors), len(os.listdir(new_directory)))
+        
+        for file in os.listdir("./repository"):
+            self.assertEqual(fnmatch.fnmatch(file, "*.txt"), True)
+        
+#------------------------------------------------
+    def test_report(self):
+        orig_stdout = sys.stdout
+        f =  open("report.dat", "w")
+        sys.stdout = f
+        mailroom.report(mailroom.donors)
+        sys.stdout = orig_stdout
+        f.close()
+        
+        with open("report.dat", "r") as f:
+            rows = [line for line in f]  
+        
+        del rows[0:2]
+        self.assertEqual(rows[1].replace(" ", ""), "DonorName|TotalGiven|NumGifts|AverageGift\n")
+        self.assertEqual(rows[3].replace(" ", ""), "Name3|$1,000,000.00|1|1,000,000.00\n")
+        self.assertEqual(rows[3].replace(" ", "").split("|")[1], "$1,000,000.00")
+ 
+#------------------------------------------------
+    def test_switch_dict(self): 
+        test = ['name_donation_letter', 'report', 'sendletter_everyone', 'quit_program']
+        switch_dict_val = [str([item for item in mailroom.switch_dict.values()][i]).split()[1] for i in range(len(mailroom.switch_dict))]
+        self.assertCountEqual(switch_dict_val, test)
+        self.assertListEqual(switch_dict_val, test)
+        self.assertEqual(switch_dict_val, test)
+        
+#------------------------------------------------
+    def test_name_donation_letter(self):
+        test_donors = {"A":[1,2], "B":[3,4]}
+        test_print_letter = "Thank you for your donation"
+        test_bad_name_entered = "We are sorry,"
+        test_negative_amount_entered = "We accept only positive donations."
+        test_wrong_amount = "Wrong amount"
+        test_error_on_wrong_ammount = "TypeError"
```

Comment:
it looks like you're testing different cases here? each one should be a separate test.

it is best for tests to be declarative and clear (referring to if statements below)